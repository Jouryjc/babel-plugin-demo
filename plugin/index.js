const { declare } = require('@babel/helper-plugin-utils');

module.exports = declare((api, options) => {
    api.assertVersion(7);
    const { types } = api;
    
    return {
        name: 'my-test-plugin',
        pre (file) {
            
        },
        visitor: {
            Identifier (path, state) {
                if (path.node.name === 'spliceText') {
                    path.node.name = 'spliceTextCopy';
                    
                    // 拿到当前 Identifier 的父节点也就是整个表达式
                    const parent = path.parent;
                    const args = parent.arguments;

                    // 只有一个参数，不需要处理
                    if (args.length === 1) {
                        return;
                    }

                    // 构建空对象的 ast
                    const params = types.objectExpression([]);

                    // 从 1 开始，遍历参数，然后塞进对象表达式的 properties 中
                    for (let i = 1, len = args.length; i < len; i++) {
                        params.properties.push(
                            types.objectProperty(
                                types.numericLiteral(i-1),
                                args[i]
                            )
                        )
                    }

                    parent.arguments.splice(1);
                    parent.arguments.push(params);
                    path.skip();
                }
            }
        },
        post (file) {

        }
    };
});