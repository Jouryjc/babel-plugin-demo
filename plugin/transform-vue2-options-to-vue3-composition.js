const { declare } = require('@babel/helper-plugin-utils');

module.exports = declare((api, options) => {
    api.assertVersion(7)
    const { types, template } = api
    const bodyContent = []

    const propertyMap = {
        props: {},
        state: {}
    }

    function assetType(property, type) {
        return types.isIdentifier(property?.key, { name: type })
    }

    /**
     * 转换props
     * options 中的 props 作为 composition api 的参数
     */
    function genDefineProps(property) {
        bodyContent.push(
            types.variableDeclaration(
                'const',
                [
                    types.variableDeclarator(
                        types.identifier('props'),
                        types.callExpression(
                            types.identifier('defineProps'),
                            [
                                property.value
                            ]
                        )
                    )
                ]
            )
        )
    }

    /**
     * 转换data
     * options api 中的 data 转成 composition api 中的 reactive 函数
     */
    function genReactiveData(property) {
        const reactiveDataAST = template.ast(`const state = reactive();`);

        reactiveDataAST.declarations[0].init.arguments.push(property.value.body.body[0].argument)

        // const toRefsAST = template.ast(`toRefs(state)`);

        bodyContent.push(
            reactiveDataAST,
            // toRefsAST
        )
    }

    /**
     * 转换computed
     * options api 中的 computed 转成 composition api 中的 computed 函数
     */
    function genComputed(property) {
        const properties = property.value.properties
        const computedArr = []

        properties.forEach(item => {
            const fnName = item.key.name

            let computedAST = null

            if (types.isObjectMethod(item)) {
                computedAST = template.ast(`const ${fnName} = computed(() => {})`)
                computedAST.declarations[0].init.arguments[0].body = item.value.body
            } else if (types.isObjectProperty(item)) {
                computedAST = template.ast(`const ${fnName} = computed()`)
                computedAST.declarations[0].init.arguments.push(item.value)
            }

            computedArr.push(computedAST)
        })

        bodyContent.push(...computedArr)
    }
    
    function genMethods (property) {
        const properties = property.value.properties
        const methodArr = []

        properties.forEach(item => {
            const fnName = item.key.name

            let ast = null
            
            if (types.isObjectProperty(item)) {
                ast = template.ast(`const ${fnName} = () => {}`)
                ast.declarations[0].init.params = item.value.params
                ast.declarations[0].init.arguments[0].body = item.value.body
            }

            methodArr.push(ast)
        })

        bodyContent.push(...methodArr)
    }

    return {
        name: 'transform-vue2-options-to-vue3-composition-plugin',
        pre(file) {
            console.log(file)
        },
        visitor: {
            // ObjectProperty (path) {
            //     console.log(path)
            // },
            // ThisExpression: {
            //     enter (path) {

            //     }
            // },
            Program: {
                exit(path) {
                    const exportDefaultDeclaration = path.node.body.filter(item => types.isExportDefaultDeclaration(item))
                    const properties = exportDefaultDeclaration[0].declaration.properties

                    properties.forEach(property => {
                        if (assetType(property, 'props')) {
                            genDefineProps(property)
                        }

                        if (assetType(property, 'data')) {
                            genReactiveData(property)
                        }

                        if (assetType(property, 'computed')) {
                            genComputed(property)
                        }

                        if (assetType(property, 'methods')) {
                            genMethods(property)
                        }
                    })

                    path.node.body.push(...bodyContent)
                }
            }
        },
        post(file) {

        }
    };
});