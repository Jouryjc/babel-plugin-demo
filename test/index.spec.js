import plugin from '../plugin/index';
import pluginTester from 'babel-plugin-tester';

pluginTester({
    plugin: plugin,
    tests: {
        'no-params': {
            code: `spliceText('有趣的灵魂')`,
            snapshot: true
        },
        'has-params': {
            code: `spliceText('我有一只小{0}，我从来都不{1}', '毛驴', '骑')`,
            snapshot: true
        }
    }
})