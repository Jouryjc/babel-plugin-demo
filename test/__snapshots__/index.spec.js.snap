// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unknown plugin has-params: has-params 1`] = `

spliceText('我有一只小{0}，我从来都不{1}', '毛驴', '骑')

      ↓ ↓ ↓ ↓ ↓ ↓

spliceTextCopy("我有一只小{0}，我从来都不{1}", {
  0: "毛驴",
  1: "骑",
});


`;

exports[`unknown plugin no-params: no-params 1`] = `

spliceText('有趣的灵魂')

      ↓ ↓ ↓ ↓ ↓ ↓

spliceTextCopy("有趣的灵魂");


`;
